sklearn/cluster/_optics.py

```diff
@@ -438,7 +438,7 @@ def _compute_core_distances_(X, neighbors, min_samples, working_memory):
 
 @validate_params(
     {
-        "X": [np.ndarray, "sparse matrix"],
+        "X": [np.ndarray, "sparse container"],
```

Comment:
@jjerphan In fact tests would fail without these changes, unless I do not change the validation for "sparse matrix". This involves some confusing issues and let me make things a bit clearer. I think there are two possibilities for this PR:

1. Let "sparse matrix" and "sparse container" have the same validation, i.e., `sparse.issparse()` instead of changing the validation of "sparse matrix" to `sparse.isspmatrix()`. In this case I can leave the parameter validations unchanged.

2. Change the validation of "sparse matrix" to `sparse.isspmatrix()` and add the option "sparse container". This is, however, much more complicated, because we would have to touch the public API to make tests pass. As a matter of fact, I would expect much more tests to fail with this change (meaning that more parameter validations need to be changed to make things pass). However, only a few tests failed for the following reason: *"array-like" is covering sparse matrices and sparse arrays, so that if a function/class accepts both "array-like" and "sparse matrix" then there will be no error. In other words, with the current validation system, ["array-like", "sparse matrix"] is the same as ["array-like"], which I do not think is reasonable.* To take this option, we may need to: (1) change "array-like" to not accept sparse stuff; (2) with the updated "sparse matrix" validation, find out all tests that are failing (I would expect a lot), and from those failing tests I think we can know which functions/classes have been tests on sparse arrays, and we can change the constraints from "sparse matrix" to "sparse container".

I'm actually not sure which option to take. Option 1 is somehow doing nothing, in which case I think we can just discard this PR. Option 2 is more reasonable, i.e., this PR is not only to extend tests but also to change the public API and fix parameter validations, but perhaps it is too much. Well, perhaps we can first make a PR to fix the problem of "array-like" covering sparse matrices and sparse arrays (if this is indeed not the desired behavior), then do the rest of Option 2 in this PR with the title changed. I do want to know what maintainers think, also @jeremiedbb.

Reply:
I think sparse containers are considered to be array-like. So option 2 works for me, but I would wait for the opinion of another maintainer.